### PR TITLE
Do not expand predefined references in Stream::consume_reference

### PR DESCRIPTION
**Pull request by Jesse-Bakker**
When processing entity declaration, only internal references should be expanded. Predefined references should be bypassed. Just parsing these as `Entity` instead of `Char` allows for this

### CODE REVIEW

**Comment by RazrFalcon** [comment]
Can you add a test case for this (I will fix ci soon)? And can you confirm that it doesn't brake [roxmltree](https://github.com/RazrFalcon/roxmltree) tests?

**Comment by Jesse-Bakker** [comment]
I added a test. There are test failures in roxmltree though, but that was to be expected. Predefined entities should be expanded when their containing entity  is referenced in the xml document.

For context, the reason for this change is that I ran into some issues with expansion of DTD entity references where the entity contained an element with a name containing character references. To fix that properly in roxmltree, I need this change, as character references are expanded in entities, but predefined entity references are not. See https://www.w3.org/TR/xml/#sec-entexpand for examples

**Comment by RazrFalcon** [comment]
Hm... Yes, this part of XML is pretty nightmarish. I've spend a lot of time on it and Im still not sure how it suppose to work.

**Comment by RazrFalcon** [comment]
Oh, this is part of the steam API, not tokenizer... Can you put those tests into the doc comment, like this:

```rust
    /// Consumes an XML reference.
    ///
    /// Consumes according to: <https://www.w3.org/TR/xml/#NT-Reference>
    ///
    /// # Errors
    ///
    /// - `InvalidReference`
    ///
    /// # Examples
    ///
    /// Escaped character:
    ///
    /// ```
    /// use xmlparser::{Stream, Reference};
    ///
    /// let mut stream = Stream::from("&#xe13;");
    /// assert_eq!(stream.consume_reference().unwrap(), Reference::Char('\u{e13}'));
    /// ```
    ///
    /// Predefined references will not be expanded.
    /// [Details.]( https://www.w3.org/TR/xml/#sec-entexpand)
    ///
    /// ```
    /// use xmlparser::{Stream, Reference};
    ///
    /// let mut stream = Stream::from("&le;");
    /// assert_eq!(stream.consume_reference().unwrap(), Reference::Entity("le"));
    /// ```
    ///
    /// Named reference:
    ///
    /// ```
    /// use xmlparser::{Stream, Reference};
    ///
    /// let mut stream = Stream::from("&other;");
    /// assert_eq!(stream.consume_reference().unwrap(), Reference::Entity("other"));
    pub fn consume_reference(&mut self) -> Result<Reference<'a>> {
        self._consume_reference().map_err(|_| StreamError::InvalidReference)
    }
```

Sorry once again. I haven't looked into this project for a while.

File: src/stream.rs [hunk 1]
```diff
@@ -346,7 +346,7 @@ impl<'a> Stream<'a> {
         Ok(())
     }
 
-    /// Consumes an XML character reference if there is one.
+    /// Consumes an XML character reference if there is one. See [try_consume_reference]
```
Comment (RazrFalcon) [inline src/stream.rs:349]: I'm sorry once again, but shouldn't it bee `See [consume_reference]`?

Actually, it should be:

```rust
/// Consumes an XML character reference if there is one.
///
/// See [`consume_reference()`] for details.
///
/// [`consume_reference()`]: #method.consume_reference
```

Comment (Jesse-Bakker) [reply src/stream.rs:349]: Ah, yes you are correct. I believe with the new intra-doc linking feature of rustdoc the explicit anchor is not necessary right?

Comment (RazrFalcon) [reply src/stream.rs:349]: This library doesn't use intra-doc linking yet. It has to be converted first. It better be done as a separate pull request, if needed.

**Comment by RazrFalcon** [comment]
Thanks again!

**Comment by rcoh** [comment]
This breaks roxmltree (or at least it causes a behavior regression). Now, when `&lt;` et al are passed to `roxmltree`, the parser will error out.

**Comment by RazrFalcon** [comment]
Wtf... I have yanked 0.13.4

**Comment by Jesse-Bakker** [comment]
This was expected to be a breaking change right?

roxmltree needs some changes to expand predefined entities itself in the right places (for example by having them in the `entities` field of the parser state by default), as right now, these references are to undefined entities.

**Comment by RazrFalcon** [comment]
I wasn't expecting it to break all entities. This was my mistake. I should have tested everything myself.

**Comment by RazrFalcon** [comment]
@Jesse-Bakker Here is how `lxml` parses your example:

```yaml
Document:
  - Element:
      tag_name: root
      children:
        - Text: "\n"
        - Element:
            tag_name: ด
            children:
              - Text: "<"
        - Text: "\n"
```

Does this mean that lxml/libxml2 is also broken? I doubt that.

**Comment by Jesse-Bakker** [comment]
No, lxml is not broken. The entity should be parsed to `<!ENTITY ent "<ด>&lt;</ด>">`. When that entity is references from the body (`<root>
&ent;
</root>`, the entity body is parsed and the `&lt;` is expanded to `<`.

If, instead of `&lt;` a character reference to `<` is used (`&#x3c`), the entity definition should fail to parse

**Comment by RazrFalcon** [comment]
Can you provide an example that should fail?

**Comment by Jesse-Bakker** [comment]
Should fail:
```xml
<!DOCTYPE root
<!ENTITY ent "<&#xE14;>&#x3c;</&#xE14;>">
]>
<root>
&ent;
</root>
```
, because `&#x3c` (`<`) will make the entity expand to `<!ENTITY ent "<ด><</ด>">` which is invalid.
```xml
<!DOCTYPE root
<!ENTITY ent "<&#xE14;>&lt;</&#xE14;>">
]>
<root>
&ent;
</root>
```
should succeed, because the entity will expand to `<!ENTITY ent "<ด>&lt;</ด>">`

**Comment by RazrFalcon** [comment]
Well, I have no idea how to fix it in roxmltree. I've created https://github.com/RazrFalcon/roxmltree/issues/50 for now. Maybe I will be able to wrap my head around it.

**Pull request merged**